Add file storage option

nucleogenesis opened the pull request:
<!--
 1. Following guidance below, replace …'s with your own words
 2. After saving the PR, tick of completed checklist items
 3. Skip checklist items that are not applicable or not necessary
 4. Delete instruction/comment blocks
-->

## Summary
<!--
 * description of the change
 * manual verification steps performed
 * screenshots if the PR affects the UI
-->

In order to allow us to use a cloud backend for the File Storage in Django, this adds a value to options.py which defaults to the Django FileSystemStorage (which... it did anyway but now we make sure of it).

This should make it configurable on BCK such that if there is a module that is a Google cloud backend class that implements the Django Storage class, then it can be added as the value for the settings.

For example, if we have a new class "GCloudStorage" in `kolibri.core.storage` then we would use that class if we set the option added here to `kolibri.core.storage.GCloudStorage`.

---

_This is very much a first whack_ -- one thing I'm not clear on is if by naming the option by the name that Django would look to in the env vars `DEFAULT_FILE_STORAGE` does the Kolibri options.py stuff automatically apply that setting because of the matching name?

## References
<!--
 * references to related issues and PRs
 * links to mockups or specs for new features
 * links to the diffs for any dependency updates, e.g. in iceqube or the perseus plugin
-->

Fixes #9441 (or at least begins to address it)

## Reviewer guidance
<!--
 * how can a reviewer test these changes?
 * are there any risky areas that deserve extra testing
-->

@pcenov - could you please test all workflows that involve:

- importing from CSV
- generating a CSV
- downloading a generated CSV

The changes I've made here should basically have no effect to the user for any of those workflows.

Once this passes regression testing, we can deploy it to a BCK instance and do final testing there.

----

## Testing checklist

- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests


## PR process

- [ ] PR has the correct target branch and milestone
- [ ] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

rtibbles commented on kolibri/utils/options.py:
```diff
@@ -359,6 +377,16 @@ def lazy_import_callback_list(value):
 
 base_option_spec = {
+    "FileStorage": {
+        "DEFAULT_FILE_STORAGE": {
```
I think I'd want to hew closer to the pattern we have for the Cache and Database options here - and offer simple, pre-specified string options that refer to specific backends. If someone really wants to run a custom backend, they can override the settings file and do what they like.

That way, with specific backends in mind, we can then explicitly enumerate the additional things that need to be specified in each case - for example the default `"file_system"` backend value will need a path, if it's a GCS backend then other things might be required (or may be automagically configured in some cases).

rtibbles commented on kolibri/utils/options.py:
```diff
@@ -271,6 +273,22 @@ def multiprocess_bool(value):
+    except ImportError:
+        logger.error("Default file storage is not available.")
+        raise VdtValueError(value)
+    except Exception:
```
Shouldn't ever be catching a bare Exception, unless for very good reason - it can hide a multitude of sins.

rtibbles commented on kolibri/utils/options.py:
```diff
@@ -271,6 +273,22 @@ def multiprocess_bool(value):
+        modules = value.split(".")
+        klass = modules.pop()
+        module_path = ".".join(modules)
+        module = importlib.import_module(module_path)
```
Note that Django exposes a utility called `import_string` which we already use in this module for loading classes by a string dot path, so this seems preferable to use here.

rtibbles reviewed (commented):
I think we probably want to keep the options.py interface restricted, others can override settings if they see fit, but we should focus on enabling the specific options we need.

Hint: if you do `kolibri configure list-env` you will see a complete list of available env vars for configuration (which will also show how your new option can be set as an env var).

nucleogenesis commented on kolibri/utils/options.py:
```diff
@@ -15,6 +15,7 @@
 from configobj import ConfigObj
 from configobj import flatten_errors
 from configobj import get_extra_values
+from django.core.files.storage import Storage
```
This is probably why flake8 wouldn't let pre-commit pass but it didn't give me useful output

jredrejo commented on kolibri/utils/options.py:
```diff
@@ -737,6 +766,7 @@ def _get_validator():
             "url_prefix": url_prefix,
             "bytes": validate_bytes,
             "multiprocess_bool": multiprocess_bool,
+            "storage_option": storage_option,
```
how will this work with database based cache?

nucleogenesis replied:
Hm - I'm not sure. I assumed this was only related to validation on initialization

rtibbles replied:
I don't see that this would affect anything to do with the cache.

nucleogenesis commented on kolibri/core/auth/management/commands/bulkexportusers.py:
```diff
@@ -203,6 +207,17 @@ def csv_file_generator(facility, filepath, overwrite=True):
+            logger.info("File saved - Path: {}".format(file_storage.path(file)))
+        except NotImplementedError:
+            # But if path is not implemented, we assume we can get the URL
+            logger.info("File saved - Path: {}".format(file_storage.url(file)))
+            logger.info("File saved - Size: {}".format(file_storage.size(file)))
```
@jredrejo I've tried several things around here, but no matter what I do I the file always shows size 0... I've confirmed that there are users (usernames is full of data, for example) -- so I'm not sure why the `writer` isn't updating the file object here...

jredrejo replied:
The problem was not in reading the file info, but writing it. It had 0 bytes. I have made a commit in this PR to ensure data is flushed into the BytesIO and now it's working:
```
INFO     2024-12-26 20:45:45,313 Invoking command bulkexportusers
INFO     2024-12-26 20:45:45,373 Creating users csv file /home/jose/.kolibri/log_export/Kolibri en casa de admin_3da4_users.csv
  [#######################################################################################################################################################################################-------------------------------------------------------------]   75%INFO     2024-12-26 20:45:46,767 File saved - Path: https://storage.googleapis.com/kdp-csv-reporting-develop/Kolibri_en_casa_de_admin_3da4_users.csv
INFO     2024-12-26 20:45:47,073 File saved - Size: 482
INFO     2024-12-26 20:45:47,073 Created csv file /home/jose/.kolibri/log_export/Kolibri en casa de admin_3da4_users.csv with 4 lines
```
(I've also rebased develop in the branch because there was already a conflict and more might appear soon)

nucleogenesis commented:
@jredrejo thank you for retargeting - I've rebased and restructured the PR a bit so the commit history is cleaner and hopefully easier to follow. Ready for code review when you can.

Also, could you try it locally w/ your gcloud credentials? I think I lost something that I had gotten working previously but cannot figure it out right now.

EDIT: I figured it out - had to setup my local adc creds per https://cloud.google.com/docs/authentication/set-up-adc-local-dev-environment

nucleogenesis commented:
@rtibbles [this failing test](https://github.com/learningequality/kolibri/actions/runs/12958418541/job/36148798528?pr=12590#step:5:2035) seems to be related to the original issue itself, possibly, because it is ultimately trying to get a hold of the DB file using `os.path.join(KOLIBRI_HOME...)` instead of the `DefaultStorage()`. 

That repair_db stuff calls to get the [default backup folder](https://github.com/learningequality/kolibri/blob/b6f59fef14672a1fea614e9d038898498936f93e/kolibri/core/deviceadmin/utils.py#L26-L27). 

The local storage uses [the MEDIA_ROOT by default](https://docs.djangoproject.com/en/5.1/ref/files/storage/#django.core.files.storage.FileSystemStorage.location).

Not sure the best path forward so would appreciate your thoughts.

nucleogenesis commented on kolibri/deployment/default/settings/base.py:
```diff
@@ -190,6 +190,14 @@
+if not os.environ.get("DEFAULT_FILE_STORAGE"):
+    if conf.OPTIONS["FileStorage"]["STORAGE_BACKEND"] == "gcs":
+        DEFAULT_FILE_STORAGE = "kolibri.utils.file_storage.KolibriFileStorage"
+        BUCKET_NAME = os.getenv("GCS_BUCKET_NAME") or "kdp-csv-reporting-develop"
```
This is borrowed from @jredrejo but should probably be set to a different fallback bucket.

We also may want to ensure that the GCS_BUCKET_NAME var is set to instance-specific buckets? 
cc @DXCanas

rtibbles replied:
We should have a BUCKET NAME option in our options.py as well - we definitely should not be hard coding anything specific about any platform in here.

If `GCS_BUCKET_NAME` is an environment variable that is set by default on GCP (but I don't think it is) we can specify specific env vars in the options configuration to draw values from.

rtibbles replied:
Rather than subclassing GoogleCloudStorage, we should just set the additional settings here for it.

nucleogenesis commented on kolibri/utils/file_storage.py:
```diff
@@ -0,0 +1,9 @@
+# https://django-storages.readthedocs.io/en/latest/backends/gcloud.html#google-cloud-storage
+
+
+class KolibriFileStorage(GoogleCloudStorage):
```
I think here we can override the [default values ](https://docs.djangoproject.com/en/5.1/ref/files/storage/#django.core.files.storage.FileSystemStorage.location) for the location if needed

rtibbles replied:
Not clear to me why this file is necessary - we can just directly use the GoogleCloudStorage class and use settings (with their own options in options.py to allow them to be defined).

https://django-storages.readthedocs.io/en/latest/backends/gcloud.html#settings

nucleogenesis commented on kolibri/core/auth/test/test_bulk_import.py:
```diff
@@ -15,8 +15,12 @@
+file_storage = DefaultStorage()
+
+
+def random_filename():
```
In hindsight, I think that maybe just cleaning up the files after each test could have avoided needing this.

nucleogenesis commented on kolibri/core/auth/test/test_bulk_import.py:
```diff
@@ -150,18 +153,18 @@ def import_exported_csv(self):
     def test_dryrun_from_export_csv(self):
-        with open_csv_for_reading(self.filepath) as source:
-            header = next(csv.reader(source, strict=True))
+        source = file_storage.open(self.filepath, "r")
```
Looking back at this, the reassignment of `source` makes this a it odd to read in a way that it wasn't when they were created inside of `with` statements.

I'll come back through and do some cleaning

rtibbles commented:
There seems to be a problem with how files are being handled that is leaving file handles open - this is why things are erroring on windows, because it is much more sensitive to files not being closed:
```
PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'D:\\a\\kolibri\\kolibri\\.pytest_kolibri_home\\media\\UserImportCommandTestCase.csv'
```

I haven't looked through the code in detail to see where this might be happening, but it suggests that something isn't being handled entirely properly.

rtibbles commented on kolibri/core/auth/csv_utils.py:
```diff
@@ -174,7 +176,7 @@ def csv_file_generator(facility, filepath, overwrite=True, demographic=False):
     )
 
-    csv_file = open_csv_for_writing(filepath)
+    csv_file = io.StringIO()
```
Why are we using `io.StringIO()` here but `io.BytesIO` below?

rtibbles commented on kolibri/utils/options.py:
```diff
@@ -385,6 +405,17 @@ def csp_source_list(value):
+            "description": """
+            The storage backend class that Django will use when managing files. The class given here must implement
+            the Django files.storage.Storage class.
+            """,
```
Let's add addiitonal options here for the ACL, bucket name, etc - we don't need to expose all the settings that are available, but for anything we need set, we can include here.

rtibbles reviewed (changes requested):
The changes here seem to have completely ignored the purpose of the two helper functions `open_csv_for_writing` and `open_csv_for_reading` - I think the diff could be significantly reduced by leaning into the existence of these two functions and consolidating the logic in there - with the small tweak as you have done that we change `filepath` to `filename`, so that it is clear it is not a literal disk file path.

We can also probably conditionalize whether we write directly to the file or to memory. Writing large CSV files to memory in a non-cloud environment is probably not going to be a great idea.

The other alternative is we could just write directly to the storage object in either case unless we know _absolutely_ that it is too slow when writing to GCS? Given it's happening in an asynchronous task, and we have to write to it eventually... I am not sure how writing to memory first necessarily helps.

rtibbles commented:
@nucleogenesis and I had a quick sync up to see how to consolidate the file handling here - seems like there are some deficiencies in the Django Storage conformance to the Python file object spec https://forum.djangoproject.com/t/file-open-to-support-different-encodings/21491 - but an idea there to wrap the file object in a `TextIOWrapper` to ensure we generate with the correct encoding and newlines seems promising!

pcenov commented:
Hi @nucleogenesis - on my end while regression testing the .csv generation I noticed that it's no longer possible to generate the same .csv file twice. For example if I go to `Facility > Data` and generate a sessions logs file for the current day, it gets generated correctly, but if I attempt to to the same thing 5 minutes later it still says "Generated 5 minutes ago."
The same is valid when generating a new user .csv file - the first time it works fine, but if I go ahead and add several new users, go back to  _Facility > Data_ and click again the "Generate new user CSV file" it will seem that it's generated correctly but when you open it you don't see the newly added users. There are no errors in the console.

Here are the logs from one of my devices though: 
[logs.zip](https://github.com/user-attachments/files/18560833/logs.zip)


https://github.com/user-attachments/assets/eb497396-a39e-4e53-8214-28ea1a15ec1d

nucleogenesis committed:
add django-storages gcs lib, filestorage options & base settings

nucleogenesis committed:
update open_for_(read|write)_csv to use default_storage w/ utf-8-sig encoding

nucleogenesis committed:
update csv open/read handling & bulk im/export tests

nucleogenesis committed:
encode open_csv_for_reading

nucleogenesis committed:
update bulkexportusers, facility views around CSVs, passing tests

rtibbles commented on kolibri/plugins/facility/views.py:
```diff
@@ -109,19 +106,16 @@ def exported_csv_info(request, facility_id):
-        if os.path.exists(log_path):
-            csv_statuses[log_type] = os.path.getmtime(log_path)
+
+        if default_storage.exists(filename):
```
According to Sentry there's a path whereby this variable doesn't exist yet: https://learningequality.sentry.io/issues/6321649781/?alert_rule_id=14557147&alert_type=issue&notification_uuid=f911561f-142a-43f1-b240-99174fc0f77d&project=1378215&referrer=slack

DXCanas replied:
Actually had a call with @nucleogenesis earlier about his access challenges; the `django-storages` library suggests using [oath2 creds for use with creds when passing settings to the library](https://django-storages.readthedocs.io/en/latest/backends/gcloud.html#settings) like so:
```
from google.oauth2 import service_account

GS_CREDENTIALS = service_account.Credentials.from_service_account_file(
    "path/to/credentials.json"
)
```
But i'd like to try just using [Application Default Credentials](https://googleapis.github.io/google-api-python-client/docs/oauth-server.html#application-default-credentials); I think that might work across local dev and cloud envs
```
import google.auth
credentials, project = google.auth.default()
```

nucleogenesis committed:
update default options; handle credentials for google.auth correctly

nucleogenesis committed:
fix dates elapsed time generation

nucleogenesis commented:
@pcenov I've made some significant changes here and I've been able to test it locally in both the Google Cloud & the local context. 

Can you go through the various CSV downloads in Kolibri again to confirm?

I will connect with you once it is testable in a cloud context if needed.

@rtibbles things are ready for re-review code-wise, I'll annotate a few areas w/ thoughts & questions as well

nucleogenesis commented on kolibri/deployment/default/settings/base.py:
```diff
@@ -190,6 +190,21 @@
+        # Initialize GS_CREDENTIALS as a proper google.auth.credentials.Credentials object
+        import google.auth
+
+        GS_CREDENTIALS, _ = google.auth.default()
```
@DXCanas this is where I ended up putting the code we need to generate the credentials object. @rtibbles is there somewhere better to put this?

rtibbles replied:
Could we initialize this during the options parsing as a second step in validation? That way we could fail much earlier if the credentials are incorrectly configured.

rtibbles commented on kolibri/deployment/default/settings/base.py:
```diff
@@ -190,6 +190,21 @@
+# File Storage Backend
+# https://docs.djangoproject.com/en/3.2/ref/files/storage/
+
+if not os.environ.get("DEFAULT_FILE_STORAGE"):
```
What's the extra environment variable here for? Can we not just gate this on the conf.OPTIONS?

rtibbles replied:
This is still unresolved, I'm not seeing any need for this extra check.

rtibbles commented on kolibri/utils/options.py:
```diff
@@ -271,6 +271,26 @@ def multiprocess_bool(value):
+    if value == "gcs":
+        try:
+            from storages.backends.gcloud import GoogleCloudStorage  # noqa
+
```
So my thought was we could try to initialize the credentials here - presumably it throws a predictable error which we can catch and raise the validation error.

rtibbles replied:
Looks like we've not completely resolved this?

nucleogenesis committed:
update storages module; allow defaults to do their magic for auth; fixes

nucleogenesis commented:
@rtibbles updating the google-storages module and removing the GS_CREDENTIALS handling altogether just worked locally <3

pcenov commented:
Hi @nucleogenesis, I confirm that the generation of the session and summary logs is working correctly now however when I try to import a user CVS file I am repeatedly getting the following error even without having modified in any way the just exported csv file:

```
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/kolibri/core/tasks/job.py", line 326, in execute
    result = func(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/kolibri/core/tasks/registry.py", line 237, in __call__
    return self.func(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/kolibri/core/auth/tasks.py", line 188, in importusersfromcsv
    delete=delete,
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/core/management/__init__.py", line 181, in call_command
    return command.execute(*args, **defaults)
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/core/management/base.py", line 398, in execute
    output = self.handle(*args, **options)
  File "/usr/lib/python3/dist-packages/kolibri/core/tasks/management/commands/base.py", line 28, in handle
    return self.handle_async(*args, **options)
  File "/usr/lib/python3/dist-packages/kolibri/core/auth/management/commands/bulkimportusers.py", line 873, in handle_async
    self.number_lines = self.get_number_lines(filepath)
  File "/usr/lib/python3/dist-packages/kolibri/core/auth/management/commands/bulkimportusers.py", line 717, in get_number_lines
    with open_csv_for_reading(filepath) as f:
  File "/usr/lib/python3.6/contextlib.py", line 81, in __enter__
    return next(self.gen)
  File "/usr/lib/python3/dist-packages/kolibri/core/utils/csv.py", line 50, in open_csv_for_reading
    with default_storage.open(filename, "rb") as f:
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/core/files/storage.py", line 38, in open
    return self._open(name, mode)
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/core/files/storage.py", line 243, in _open
    return File(open(self.path(name), mode))
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/core/files/storage.py", line 338, in path
    return safe_join(self.location, name)
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/utils/_os.py", line 31, in safe_join
    'component ({})'.format(final_path, base_path))
django.core.exceptions.SuspiciousFileOperation: The joined path (/home/osboxes/.kolibri/temp/tmpv3rj_54v.upload.csv) is located outside of the base path component (/home/osboxes/.kolibri/media)
```
Here's a video as well:

https://github.com/user-attachments/assets/a1f7378c-3086-461c-810a-74a76690883c

Logs: [logs.zip](https://github.com/user-attachments/files/19028367/logs.zip)

nucleogenesis committed:
add logging to view for debugging in cloud

nucleogenesis commented on kolibri/utils/options.py:
```diff
@@ -385,6 +405,45 @@ def csp_source_list(value):
 
 
 base_option_spec = {
+    "FileStorage": {
```
.

nucleogenesis committed:
ImportUsersFromCSVValidator: Use default storage for uploaded file

nucleogenesis commented:
@pcenov I've updated the code so that you should be able to import a user by CSV. I've tested in the cloud env, can you confirm it is resolved locally as well?

pcenov commented:
Hi @nucleogenesis - I confirm that I can import users now, however for some reason nothing happens when I click the link 'Generate new user CSV file':

https://github.com/user-attachments/assets/1db852b7-b7a6-4d63-a061-7b417f99c580

[logs.zip](https://github.com/user-attachments/files/19428560/logs.zip)

nucleogenesis committed:
When opening CSV for writing, overwriting is implied

In the task we have an overwrite flag that defaults to True; but when it
is toggled to False it basically just serves as a way of avoiding
overwriting by mistake because we just throw an error about it if
overwrite=False and the file already exists

nucleogenesis commented:
@pcenov thank you for catching that - I've resolved the issue and tested it in cloud & local envs. 

@rtibbles in https://github.com/learningequality/kolibri/pull/12590/commits/15838260d102bb891968f0ae762243bf5b4292ee I removed some code I added in hopes of allowing for non-overwrite operations.

I ran into errors where it said the file wasn't open for writing... so I changed rb+ to wb+ and got "file not open for writing" and I got to wondering if the wrapping & encoding of a file for write+read was causing the issues but I couldn't work out anything to substantiate that.

In any case - the bulkuserexport's overwrite flag is simply a way to exit the task early if the file already exists to avoid overwriting, so it seems fair for `open_csv_for_writing` to always overwrite when given the same filename.

Just wanted to get your thoughts

pcenov commented:
@nucleogenesis - I wanted to verify that the issue is fixed but it's still the same in the latest build asset.

nucleogenesis committed:
upgrade routine to cleanup old export log & csv generation files

nucleogenesis committed:
prefix CSV filename strings w/ appropriate subdir name for their context

nucleogenesis committed:
handle removal of user import temp files

dryrun files need to stick around if they pass so it's still there for
the wetrun so we delete on exception or if we're on a wetrun

nucleogenesis committed:
update csv cleanup task & test tow orkw ith log_export dir

rtibbles commented on kolibri/plugins/facility/assets/src/modules/manageCSV/actions.js:
```diff
@@ -79,12 +79,12 @@ function getExportedCSVsInfo(store) {
     const data = response.data;
     if (data.session != null) {
       getCSVLogRequest(store, 'session', store.rootGetters.activeFacilityId);
-      const sessionTimeStamp = new Date(data.session * 1000);
```
What's the reason for these changes?

nucleogenesis replied:
In the facility.views we used to get the time with `os.path.getmtime(log_path)` but I used the Storage class's `get_modified_time(log_path).isoformat()`. I saw that same `.isoformat()` being used elsewhere in `exported_csv_info` and it jived well when i removed the multiplication here.

nucleogenesis commented on kolibri/core/logger/test/test_api.py:
```diff
@@ -64,7 +64,7 @@ def setUpTestData(cls):
     def test_csv_download(self):
         expected_count = ContentSummaryLog.objects.count()
-        _, filepath = tempfile.mkstemp(suffix=".csv")
+        filepath = default_storage.path("{}.csv".format(uuid.uuid4()))
```
Thanks copilot, this is a test so it won't be run in a cloud configuration

rtibbles commented on kolibri/core/auth/management/commands/exportusers.py:
```diff
@@ -57,15 +56,13 @@ def handle_async(self, *args, **options):
         else:
             filename = options["output_file"]
 
-        filepath = os.path.join(os.getcwd(), filename)
```
There is a slight issue with making this change universally in the management commands.

Our documentation has explicit instruction on using these commands to directly output a file to the current working directory, and read from the current working directory too: https://kolibri.readthedocs.io/en/latest/manage/command_line.html#import-and-export-user-data-from-a-csv-file

The completely clean approach would be separating out a lot of this functionality out of the management commands in order to allow this to be done cleanly - but in the short term, we might need to just have a way to bypass storage use and work directly with a file path to continue to support this?

rtibbles reviewed (commented):
Still some unresolved comments for options and settings, some questions, and a regrettably late in the game realization about the management commands.

I think we could spin out the management command rework into a follow up issue so this doesn't continue to linger, but it would have to be fixed before release to avoid regression.

nucleogenesis committed:
file storage settings for cloud env applied in KolibriCoreConfig#ready

nucleogenesis committed:
improved sanity check; revert settings to settings file

nucleogenesis committed:
remove old export/import users & tests; add message w/ link to docs

nucleogenesis committed:
csv open/write utils handle local & default_storage based on command params

nucleogenesis committed:
handle local vs storage in bulkexportusers based on user-given params

nucleogenesis committed:
update bulk import w/ use_storage param; apply it in relevant tasks

nucleogenesis committed:
update session & summary logs to work w/ django storage or local paths

nucleogenesis committed:
text fixes for facility downloada api and logger task api

nucleogenesis commented:
@pcenov I've updated this to make sure the CLI export/import works as well.

Could you please test the local filesystem regular CSV file export & import as you had previously for regressions. I have tested it locally and in the cloud environment successfully.

In addition to this, could you please test out the following CLI commands (note that if you put `-h` at the end, it will explain the usage):

- `kolibri manage exportlogs` - This exports Summary & Session logs. Note the `-l` flag which determines which of these you're running. Additionally, please test with the `-O` output file for the export, putting the file in a few places on your filesystem. 
- `kolibri manage bulkuserexport` - This will export all of the default facility's users to a CSV in the current working directory (where you are in your terminal). You should be able to export to a particular file using the `-O` flag. 
- `kolibri manage bulkuserimport` - This will import a CSV - please test valid and invalid CSVs and try the `--dryrun` flag, which will only validate the data.
- For **export commands** - please test them with the `-s` flag. This is newly added and should result in your files being places into your `$KOLIBRI_HOME/media/` folder (or a subfolder of it).

In general **do not worry about** testing flags that aren't related to files - whether reading or writing them. Things like overwrite, errorlines, etc are not crucial here.

Thank you in advance - I've tried these all out locally so I'm hoping you see the same!

pcenov commented:
Hi @nucleogenesis - unfortunately it seems that there is a problem with the build as it says that linting is failing. Tried rerunning the jobs but it keeps failing.

Testing with the available [build asset](https://github.com/learningequality/kolibri/suites/36502266953/artifacts/2856564508) is not possible because both the generation of new session/summary logs and new user csv file is failing:

https://github.com/user-attachments/assets/ebd0b105-7d95-48cc-8a77-9fca69a6a2fe

nucleogenesis commented:
@pcenov :thinking: 

I downloaded the assets and I'm able to generate & import/export the files. Our version hashes are the same so we're definitely on the same asset.

Could you share your logs?

rtibbles commented on kolibri/core/logger/management/commands/exportlogs.py:
```diff
@@ -127,6 +144,17 @@ def handle_async(self, *args, **options):
             else:
```
I think this if statement is now redundant? If we are doing changes completely in parallel to bulk export users we should remove this if/else statement completely and just have:
```
filename = log_info["filename"].format(facility.name, facility.id[:4], start_date[:10], end_date[:10])
```

rtibbles reviewed (commented):
Spotted one small difference between export logs and export users implementations.

rtibbles commented:
@nucleogenesis I did a quick manual test here by generating a users CSV, downloading it, and then adding a new user, and generating a new users CSV.

When I went to download this new users CSV file it did not contain the new user at all, but looking in my media folder, I do see another CSV file, so it seems that it has generated a new file rather than overwriting the existing one.

I assume this is probably what is happening in @pcenov's case as well? The logs probably won't show anything out of the ordinary, because no error has occurred, but the exported CSV files are not overwriting the previous versions.

rtibbles commented on kolibri/core/utils/csv.py:
```diff
@@ -1,14 +1,67 @@
+            )
+            yield encoded_fh
+            encoded_fh.flush()
+            default_storage.save(storage_filepath, f)
```
Looks like before we write here, we need to do an exists check and delete if so?

```
if default_storage.exists(storage_filepath):
    default_storage.delete(storage_filepath)
```
then do the save.

pcenov commented:
@nucleogenesis - yes, there are no errors in the logs as mentioned by Richard but adding them just in case:
[logs.zip](https://github.com/user-attachments/files/19564637/logs.zip)

nucleogenesis committed:
redundant conditional removed in exportlogs

nucleogenesis committed:
remove existing file before save

nucleogenesis commented:
Thanks @rtibbles for testing!

I've updated the code and tested the issues reported and things seem to be working. The elapsed time is reporting correctly and I have generated a CSV, updated it, imported it, then downloaded it again and saw the user I added.

@pcenov please give it another shot when you can

pcenov commented:
Hi @nucleogenesis - Happy to let you know that this time I didn't observe any regressions while testing the .csv file export & import and that newly generated session/summary/user csv files are correctly replacing the previously existing files.

Also no issues observed while testing the following CLI commands:

`kolibri manage exportlogs` - I confirm that it exports correctly the summary/session logs as specified. Here's an example of a working command with multiple flags: `kolibri manage exportlogs -O session_logs.csv -l session --start_date 2025-04-02T00:00:00 --end_date 2025-04-03T23:59:59 --facility 618d5168ae7eeaf1ab6c448c92c4e7e9 -w`
`kolibri manage bulkexportusers` - no issues observed while exporting with or without flags, the new -s flag is also working as specified
`kolibri manage bulkimportusers` - works as specified

Something that's probably known to everyone involved but mentioning it just in case is that the documentation at https://kolibri.readthedocs.io/en/latest/manage/command_line.html will have to be updated to include the new changes.

Also the following commands listed there are not working:

`kolibri manage exportlogs --log-type summary`
`kolibri manage exportlogs --log-type session`

nucleogenesis committed:
Bypass a test that fails only on Windows in facility..test_api

rtibbles reviewed (approved):
All comments addressed, manual QA approved. Only change since QA approval has been to tweak a failing test on Windows that seems to be due to a difficult to trace test interaction.
